### PR TITLE
Decentralize includes

### PR DESCRIPTION
**Pull request by danielhlarkin**
The purpose of this PR is to decentralize some of our header includes (i.e. reducing the number of headers included by commonly-used headers), with the end goal of reducing unnecessary recompilation. I did an analysis to determine how many other source code files depended either directly or indirectly on a given header. This PR reduces the sum of these counts by 15-20%. One of the key tools here was to use `iwyu_tool`, a script distributed with `iwyu` that plugs into a `cmake` build directory to gather the necessary information to analyze a target source file. This allows for manual auditing of individual files to try a) reduce unnecessary includes and b) include shadow dependencies which lead to breakage when changing other headers.

There is a companion enterprise PR: https://github.com/arangodb/enterprise/pull/284

Most recent Jenkins run: http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/5554/

### CODE REVIEW

File: lib/Basics/ReadWriteSpinLock.h [hunk 1]
```diff
@@ -29,6 +29,7 @@
 #include "Basics/SharedAtomic.h"
 #include "Basics/SharedCounter.h"
```
Comment (jsteemann) [inline lib/Basics/ReadWriteSpinLock.h]: I think `SharedAtomic` and `SharedCounter` are not needed here...?!

File: lib/Basics/files.cpp [hunk 2]
```diff
@@ -21,37 +21,29 @@
...
+#include <utility>
+
+#include "Basics/operating-system.h"
 
 #ifdef _WIN32
 #include <Shlwapi.h>
 #include <tchar.h>
-#include <chrono>
-#include <thread>
+#include <windows.h>
```
Comment (dothebart) [inline lib/Basics/files.cpp]: do we really require windows.h?

File: lib/Basics/memory.cpp [hunk 3]
```diff
@@ -21,12 +21,20 @@
...
 #endif
 
+#include <cstdlib>
+
+#include "Basics/Common.h"
+#include "Basics/error.h"
+#include "Basics/voc-errors.h"
+
+#ifdef _WIN32
+#include <windows.h>
```
Comment (dothebart) [inline lib/Basics/memory.cpp]: do we really need windows.h?

File: lib/Basics/operating-system.h [hunk 4]
```diff
@@ -705,9 +702,10 @@
 
 //#define _CRT_SECURE_CPP_OVERLOAD_STANDARD_NAMES     1
 
-#include <WinSock2.h>
+#include <WinSock2.h>  // must be before windows.h
 #include <io.h>
 #include <stdio.h>
+#include <windows.h>
```
Comment (dothebart) [inline lib/Basics/operating-system.h]: do we really need windows.h? at least in this place I strongly dislike it.

File: lib/SimpleHttpClient/ClientConnection.h [hunk 5]
```diff
@@ -24,12 +24,16 @@
 #ifndef ARANGODB_SIMPLE_HTTP_CLIENT_CLIENT_CONNECTION_H
 #define ARANGODB_SIMPLE_HTTP_CLIENT_CLIENT_CONNECTION_H 1
 
-#include "Basics/Common.h"
+#include <stddef.h>
+#include <memory>
 
-#include "Basics/socket-utils.h"
```
Comment (dothebart) [inline lib/SimpleHttpClient/ClientConnection.h]: <3

File: lib/V8/v8-globals.h [hunk 6]
```diff
@@ -24,17 +24,22 @@
...
 #define ARANGODB_V8_V8__GLOBALS_H 1
 
-#include "Basics/Common.h"
+#include <string.h>
+#include <cstdint>
+#include <memory>
+#include <unordered_map>
+#include <utility>
+
+#include <v8.h>
```
Comment (dothebart) [inline lib/V8/v8-globals.h]: if you include v8 here, you can skip it from V8PlatformFeature (most probably)

File: lib/V8/v8-json.ll [hunk 7]
```diff
@@ -26,11 +26,19 @@
...
+#include <cmath>
+
 #include "Basics/Common.h"
+#include "Basics/operating-system.h"
 #include "Basics/tri-strings.h"
 #include "V8/v8-json.h"
 #include "V8/v8-globals.h"
 
+#ifdef _WIN32
+#include "Basics/win-utils.h"
```
Comment (dothebart) [inline lib/V8/v8-json.ll]: Why this?

File: lib/Basics/ReadWriteSpinLock.h (hunk 1 above)
Comment (mpoeter) [reply lib/Basics/ReadWriteSpinLock.h]: Yes, they were used in the old `ReadWriteSpinLock` implementation, but this is no longer the case.

File: arangod/Agency/Job.h [hunk 8]
```diff
@@ -24,20 +24,29 @@
...
 #include <velocypack/Iterator.h>
 #include <velocypack/Slice.h>
 #include <velocypack/velocypack-aliases.h>
 
-#include <string>
+#include "Agency/AgencyCommon.h"
+#include "Agency/AgentInterface.h"
+#include "Basics/Result.h"
+#include "Basics/debugging.h"
+#include "Logger/LogMacros.h"
```
[the remark this answers is not in the document]
Comment (danielhlarkin) [reply arangod/Agency/Job.h]: Technically, no. `Logger/LogMacros.h` includes both `Logger/Logger.h` and `Logger/LoggerStream.h`. I was erring on the side of adding transitive includes as reported by iwyu. We could simply combine the three headers into one if we want to cut down on the includes. Thoughts?

File: arangod/Aql/AstNode.h [hunk 9]
```diff
@@ -24,22 +24,33 @@
 #ifndef ARANGOD_AQL_AST_NODE_H
 #define ARANGOD_AQL_AST_NODE_H 1
 
-#include "Basics/AttributeNameParser.h"
-#include "Basics/Common.h"
+#include <string.h>
```
[the remark this answers is not in the document]
Comment (danielhlarkin) [reply arangod/Aql/AstNode.h]: Yes, these are all used directly in the header. Previously they were transitively included, now it's just explicit.

File: arangod/Aql/tokens.ll [hunk 10]
```diff
@@ -23,8 +23,13 @@
 
 %{
 #include "Basics/Common.h"
-#include "Basics/conversions.h"
 #include "Basics/NumberUtils.h"
+#include "Basics/conversions.h"
+#include "Basics/operating-system.h"
+
+#if _WIN32
+#include "Basics/win-utils.h"
```
[the remark this answers is not in the document]
Comment (danielhlarkin) [reply arangod/Aql/tokens.ll]: It's necessary for the `isatty` (overwritten by a define in `Basics/operating-system.h` as `_cyg_isatty`, which is declared in `Basics/win-utils.h`).

File: arangod/Aql/ExecutionBlock.h [hunk 11]
```diff
@@ -26,11 +26,12 @@
 
 #include "Aql/AqlItemBlock.h"
 #include "Aql/BlockCollector.h"
-#include "Aql/ExecutionNode.h"
 #include "Aql/ExecutionEngine.h"
+#include "Aql/ExecutionNode.h"
 #include "Aql/ExecutionState.h"
 #include "Aql/ExecutionStats.h"
 #include "Aql/Variable.h"
+#include "Basics/system-functions.h"
```
[the remark this answers is not in the document]
Comment (danielhlarkin) [reply arangod/Aql/ExecutionBlock.h]: It's needed for `TRI_microtime()`. We'll be able to remove a lot of includes from headers if we end up enabling LTO and moving code into cpp files.

File: lib/V8/v8-json.ll (hunk 7 above)
Comment (danielhlarkin) [reply lib/V8/v8-json.ll]: To be honest, I don't remember off the top of my head, but it was to fill a dependency from some macro. We could stand to do a lot of restructuring to better abstract away the OS-dependent implementations, but that's for another day.

File: lib/V8/v8-globals.h (hunk 6 above)
Comment (danielhlarkin) [reply lib/V8/v8-globals.h]: Better to just include it both places it's used. Keeps things more self-contained and reduces risk of further breakage in the future.

File: lib/Basics/memory.cpp (hunk 3 above)
Comment (danielhlarkin) [reply lib/Basics/memory.cpp]: It is needed here for `SecureZeroMemory`.

File: lib/Basics/operating-system.h (hunk 4 above)
Comment (danielhlarkin) [reply lib/Basics/operating-system.h]: It seems you are correct and it is unneeded here.

**Pull request merged**